server/src/main/java/io/spine/server/ServerEnvironment.java

```diff
@@ -24,17 +24,29 @@
 import org.checkerframework.checker.nullness.qual.Nullable;
 
 import java.util.Optional;
+import java.util.function.Supplier;
 
-@SuppressWarnings("AccessOfSystemProperties") // OK as we need system properties for this class.
-public class ServerEnvironment {
+import static com.google.common.base.Strings.emptyToNull;
+import static java.lang.ThreadLocal.withInitial;
+
+/**
+ * The server conditions and configuration under which the application operates.
+ */
+public final class ServerEnvironment {
+
+    private static final ServerEnvironment INSTANCE = new ServerEnvironment();
 
     /** The key of the Google AppEngine runtime version system property. */
-    @VisibleForTesting
-    static final String ENV_KEY_APP_ENGINE_RUNTIME_VERSION = "com.google.appengine.runtime.version";
+    private static final String ENV_KEY_APP_ENGINE_RUNTIME_VERSION =
+            "com.google.appengine.runtime.version";
 
     /** If set, contains the version of AppEngine obtained from the system property. */
+    @SuppressWarnings("AccessOfSystemProperties") /*  Based on system property. */
     private static final @Nullable String appEngineRuntimeVersion =
-            System.getProperty(ENV_KEY_APP_ENGINE_RUNTIME_VERSION);
+            emptyToNull(System.getProperty(ENV_KEY_APP_ENGINE_RUNTIME_VERSION));
+
+    private static final ThreadLocal<DeploymentType> provider =
```

Comment:
The deployment type is the same for all threads. So using `ThreadLocal` doesn't make much sense.